### PR TITLE
Open Roles & Clean buildCreateContractTx Request

### PR DESCRIPTION
**Pull request by bjornkihlberg**
This PR fulfills 2 issues : 
- Completely #123
- Partially #68 (only the Request part of the endpoint)  

## Summary by CodeRabbit

- **New Features**
  - Search functionality now available with a new search bar.
  - Enhanced contract creation flexibility with broader input scope.
  - New dynamic type guards for JSON schemas in the Runtime Rest API.
  - Additional role configuration options for system contracts.
  - Improved transaction creation process for contract interactions.

- **Enhancements**
  - Updated documentation comments for better clarity on endpoints.
  - Streamlined import map structure for improved maintainability.

- **Bug Fixes**
  - Fixed wallet name matching logic to be case-insensitive.

- **Refactor**
  - Renamed and restructured several internal entities to align with updated functionality.

- **Style**
  - Adjusted quotes in script element creation for consistency.

- **Documentation**
  - Updated category tags in documentation comments for clearer navigation.

- **Tests**
  - Expanded test coverage to include new features and changes.

- **Chores**
  - General codebase maintenance and cleanup.

### CODE REVIEW

**Comment by nhenin** [comment]
JIRA Dependencies : PLT-8261

File: packages/adapter/src/codec.ts [hunk 1]
```diff
@@ -5,6 +5,26 @@ import JSONbigint from "json-bigint";
...
+ * }
+ * ```
+ */
+export class UnexpectedRuntimeResponse extends Error {
+  public type = "UnexpectedRuntimeResponse" as const;
+  constructor(message: string) {
+    super();
+  }
+}
+
```
Comment (bjornkihlberg) [inline packages/adapter/src/codec.ts:27]: Did you forget to pass along message?

This is more ideomatic:
```ts
class UnexpectedRuntimeResponse extends Error {
  constructor(message?: string) {
    super(message);

    Object.setPrototypeOf(this, UnexpectedRuntimeResponse.prototype);
  }
}
```

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 2]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
+   *
+   * If the user fulfills the conditions of the guarantee, and a contract gets executed, the collateral is safe.
+   * </p>
+   * @see
+   * https://docs.cardano.org/smart-contracts/plutus/collateral-mechanism
    */
   collateralUTxOs?: TxOutRef[];
+
+  /**
+   * A Marlowe Contract or a Merkleized One (referred by its source Id) to create over Cardano
```
Comment (bjornkihlberg) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:275]: Maybe "Merkleized One" should be "merkleized one" to not seem like it's a thing or a product.

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 3]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
+   *
+   * <h4>Justification</h4>
+   * <p>Creating a Marlowe Contracts over Cardano is about creating UTxO entries on the Ledger.
+   *
+   * To protect the ledger from growing beyond a certain size that will cost too much to maintain,
+   * a constraint called "Minimum ada value requirement (mininmumLovelaceUTxODeposit)" that adjust
+   * the value (in ADA) of each UTxO has been added.
+   *
+   * The more the UTxOs entries are big in size, the more the value of minimum
+   * of ADAs needs to be contained.</p>
```
Comment (bjornkihlberg) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:334]: This is slightly confounding. Shouldn't it be something along the lines of "The more UTxOs in a transaction, the higher minimum amount of ADA needs to be included."

File: packages/runtime/lifecycle/src/api.ts [hunk 4]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
+     *
+     *  1. **By Adressses** : Addresses are directly defined within the Marlowe Contract and no configuration are necessary in that context.
+     *  2. **By Roles** : Defined by {@link @marlowe.io/language-core-v1!index.RoleName | RoleNames} within the Marlowe Contract, they have to match a Token Name when created in Cardano. This field `rolesConfiguration` is about configuring this use case
+     * </p>
+     *
+     * <h4>Configuration Options</h4>
+     * <p>
+     *  
+     * - **When to create (mint)**
+     *   - **Within the Runtime** : At the contrat creation, these defined Roles Tokens will be minted "on the fly" by the runtime.
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:76]: "contrat" should be "contract"

File: packages/runtime/lifecycle/src/api.ts [hunk 5]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
+     *   - **Without the Runtime** : before the creation, these Role Tokens are already defined (via an NFT platform, `cardano-cli`, another Marlowe Contract Created, etc.. )
+     * - **How to distribute**
+     *   - **Closedly** (Closed Roles) : At the creation of contract or before, the Role Tokens are released to the participants. All the participants are knowned at the creation and therfore we consider the participation as being closed.
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:79]: "knowned" should be "known"

File: packages/runtime/lifecycle/src/api.ts [hunk 6]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
+   */
   tags?: Tags;
+  /**
+   * Cardano Metadata about the contract creation.
+   * <h4>Properties</h4>
+   * <p>
+   * Metadata can be expressed as a JSON object with some restrictions:
+   *   - All top-level keys must be integers between 0 and 2^64 - 1.
+   *   - Each metadata value is tagged with its type.
+   *   - Strings must be at most 64 bytes when UTF-8 is encoded.
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:176]: I'm assuming it should be "Strings must be at most 64 characters long when UTF-8 is encoded."

File: packages/runtime/lifecycle/src/api.ts [hunk 7]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
+   *   - Bytestrings are hex-encoded, with a maximum length of 64 bytes.
+   *
+   * Metadata aren't stored as JSON on the Cardano blockchain but are instead stored using a compact binary encoding (CBOR).
+   * The binary encoding of metadata values supports three simple types:
+   *    - Integers in the range `-(2^64 - 1)` to `2^64 - 1`
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:181]: If it's a 65-bit integer (Which sounds odd) it should be `-(2^64)` to `2^64 - 1`. I'm assuming it should be `-(2^63)` to `2^63 - 1`.

File: packages/runtime/lifecycle/src/api.ts [hunk 8]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
+
+  /**
+   * The Marlowe Runtime utilizes this Optional field to set a stake address
+   * where to send staking rewards for the Marlowe script outputs of this contract.
+   */
+  stakeAddress?: StakeAddressBech32;
+  /**
+   * @experimental
+   * Thread Roles are a details of implementation within the runtime. It allows provide a custom name
+   * if the thread role name is conflicting with other role names used.
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:55]: I'd change this to "The Thread Roles capability is an implementation details of the runtime. It allows you to provide a custom name if the thread role name is conflicting with other role names used."

File: packages/runtime/lifecycle/src/api.ts [hunk 9]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
...
-  minUTxODeposit?: number;
-};
 
-export const minUTxODepositDefault: number = 3_000_000;
+  /**
+   * Minimum Lovelace value to add on the UTxO created (Representing the Marlowe Contract Created on the ledger).This value
+   * is computed automatically within the Runtime, so this parameter is only necessary if you need some custom adjustment.
+   *
+   * <h4>Justification</h4>
+   * <p>Creating a Marlowe Contracts over Cardano is about creating UTxO entries on the Ledger.
```
Comment (bjornkihlberg) [inline packages/runtime/lifecycle/src/api.ts:199]: "Creating a Marlowe Contracts"
should either be
"Creating Marlowe contracts"
or
"Creating a Marlowe contract"

File: packages/adapter/src/codec.ts (hunk 1 above)
Comment (nhenin) [reply packages/adapter/src/codec.ts:27]: this something @hrajchert and I worked together on the side during a video call , I should remove it , Thx for spotting it and provided feedback on it !

File: jsdelivr-npm-importmap.js [hunk 10]
```diff
@@ -1,73 +1,43 @@
...
-      "https://cdn.jsdelivr.net/npm/@marlowe.io/runtime-lifecycle@0.2.0-beta/dist/bundled/esm/browser.js",
-    "@marlowe.io/runtime-lifecycle/generic":
-      "https://cdn.jsdelivr.net/npm/@marlowe.io/runtime-lifecycle@0.2.0-beta/dist/bundled/esm/generic.js",
-    "@marlowe.io/marlowe-object":
-      "https://cdn.jsdelivr.net/npm/@marlowe.io/marlowe-object@0.2.0-beta/dist/bundled/esm/marlowe-object.js",
-    "@marlowe.io/marlowe-object/guards":
-      "https://cdn.jsdelivr.net/npm/@marlowe.io/marlowe-object@0.2.0-beta/dist/bundled/esm/guards.js",
-    "lucid-cardano": "https://unpkg.com/lucid-cardano@0.10.7/web/mod.js",
-  },
+  "imports": {
```
Comment (hrajchert) [inline jsdelivr-npm-importmap.js:2]: this amount of change indicates me that you have not done a treefmt. We need to add that as a CI check. I created this Issue to remind ourselves 

https://github.com/input-output-hk/marlowe-ts-sdk/issues/115

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 11]
```diff
@@ -192,12 +199,30 @@ export const GetContractsResponseGuard = assertGuardEqual(
...
 /**
- * Request options for the {@link index.RestClient#createContract | Create contract } endpoint
- * @category Endpoints
+ * Either a Non Merkleized Marlowe Contract or a Merkleized One
+ * @category Endpoint : Build Create Contract Tx
+ */
+export type ContractOrSourceId = Contract | SourceId;
+
+/**
+ * Guard for ContractOrSourceId type
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:208]: Notice that both in the core-v1 and marlowe-objects package I document all guards [this way](https://github.com/input-output-hk/marlowe-ts-sdk/blob/main/packages/language/core/v1/src/payee.ts#L41) That allows me to eventually change All guards documentation by a simple search and replace if we decide that there is a better message.

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 12]
```diff
@@ -192,12 +199,30 @@ export const GetContractsResponseGuard = assertGuardEqual(
...
+ * Request options for the {@link index.RestClient#buildCreateContractTx | Build Create Contract Tx } endpoint
+ * @category Endpoint : Build Create Contract Tx
  */
-export interface CreateContractRequest {
-  // FIXME: create ticket to add stake address
-  // stakeAddress: void;
+export interface BuildCreateContractTxRequest {
+  /**
+   * The Marlowe Runtime utilizes this Optional field to set a stake address
+   * where to send staking rewards for the Marlowe script outputs of this contract.
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:223]: I don't think this sentence is grammatically correct. For PRs that rely heavily on documentation, we can ask @sfoo-iohk to pitch in.

My 2 cents for this is (if I understand the option correctly):

`Marlowe contracts can have staking rewards for the ADA locked in the contract. Use this field to set the recipient address of those rewards`

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 13]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
+   * @see
+   * https://docs.cardano.org/smart-contracts/plutus/collateral-mechanism
    */
   collateralUTxOs?: TxOutRef[];
+
+  /**
+   * A Marlowe Contract or a Merkleized One (referred by its source Id) to create over Cardano
+   * @see Large/Deep Contracts Support (Contract Merkleization) and `@marlowe.io/language-core`
+   */
+  contractOrSourceId: ContractOrSourceId;
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:278]: I think this is an unnecesary breaking change that makes the DX less readable.

We could split `BuildCreateContractTxRequest` into the shared parts and do something like

```
interface BuildCreateContractTxRequestShared {
  //... 
} 

interface BuildCreateContractTxRequestContract extends BuildCreateContractTxRequestShared {
  contract: Contract;
}

interface BuildCreateContractTxRequestSourceId extends BuildCreateContractTxRequestShared {
  sourceId: SourceId;
}

type BuildCreateContractTxRequest = 
  BuildCreateContractTxRequestSourceId 
| BuildCreateContractTxRequestContract
```

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 14]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
-  //       why it is required, who deposits it, and how and when you get it back.
-  minUTxODeposit: number;
+  mininmumLovelaceUTxODeposit?: number;
 
-  // TODO: Comment this and improve the generated type (currently `string | {}`)
-  roles?: RolesConfig;
   /**
-   * An optional object of tags where the **key** is the tag name (`string`) and the **value** is the tag content (`any`)
+   * @experimental
+   * Thread Roles are a details of implementation within the runtime. It allows provide a custom name
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:342]:
```
Thread Roles is an implementation detail on how the runtime...
```
We should probably include an external documentation reference. Maybe this one https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe-runtime/doc/open-roles.md

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 15]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
+   * if the thread role name is conflicting with other role names used.
+   * @default
+   *  - the Thread Role Name is "" by default.
    */
-  tags?: Tags;
+  threadRoleName?: RoleName;
+
+  /**
+   * Role Token Configuration for the contract passed in the `contractOrSourceId` field.
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:350]: Role token Configuration for the new contract.

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts [hunk 16]
```diff
@@ -211,52 +236,217 @@ export interface CreateContractRequest {
...
+   *
+   * <h4>Prerequisite</h4>
+   * <p>
+   * Participants ({@link @marlowe.io/language-core-v1!index.Party | Party}) in a Marlowe Contract can be expressed in 2 ways:
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/endpoints/collection.ts:354]: More than Prerequisite

<h4>Participants</h4>
A Marlowe contract can have two types of {@link Participants}:
1. Address: When an address is fixed in the contract we don't need to provide further configuration
2. Role: When the participation is done through a Role Token, we need to define if that token is minted as part of the contract creation transaction or if it was minted beforehand.

File: packages/runtime/client/rest/src/contract/rolesConfigurations.ts [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+  t.string,
+  (s): s is t.Branded<string, AddressBech32Brand> => true,
+  "AddressBech32"
+);
+
+/**
+ *  Cardano Address in a Bech32 format
+ *  @category Roles Configuration
+ */
+export type AddressBech32 = t.TypeOf<typeof AddressBech32Guard>;
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/rolesConfigurations.ts:24]: I think we created this in a pair session, but the idea is to replace the runtime-core one with this one and not to create yet another AddressBech32 type

File: packages/runtime/client/rest/src/contract/rolesConfigurations.ts [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+ *  @category Roles Configuration
+ */
+export const openRole = "OpenRole";
+
+/**
+ *  Definition of the Openness of a Role Token
+ *  @category Roles Configuration
+ */
+export type Openness = ClosedRole | OpenRole;
+export const OpenessGuard: t.Type<Openness, string> = t.union([
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/contract/rolesConfigurations.ts:53]: guards needs category to render properly

File: packages/runtime/client/rest/src/index.ts [hunk 19]
```diff
@@ -244,14 +235,18 @@ export function mkRestClient(baseURL: string): RestClient {
     getContractById(contractId) {
       return unsafeTaskEither(Contract.getViaAxios(axiosInstance)(contractId));
     },
-    createContract(request) {
+    buildCreateContractTx(request) {
```
Comment (hrajchert) [inline packages/runtime/client/rest/src/index.ts:238]: Also @jhbertra, can you make this name change in the backend for a future release so we maintain the 1-1 parity?

File: packages/runtime/lifecycle/src/api.ts [hunk 20]
```diff
@@ -28,14 +38,177 @@ export type RuntimeLifecycle = {
 export type ContractsDI = WalletDI & RestDI;
 
 export type CreateContractRequest = {
-  contract: Contract;
-  roles?: RolesConfig;
+  /**
+   * A Marlowe Contract or a Merkleized One (referred by its source Id) to create over Cardano
+   * @see Large/Deep Contracts Support (Contract Merkleization) and `@marlowe.io/language-core`
+   */
+  contractOrSourceId: ContractOrSourceId;
```
Comment (hrajchert) [inline packages/runtime/lifecycle/src/api.ts:45]: Same comment than the rest-client one.
cc @bjornkihlberg

**Review by hrajchert — CHANGES_REQUESTED** [review]
It is looking good, some changes are suggested

File: packages/language/core/v1/src/address.ts [hunk 21]
```diff
@@ -1,6 +1,5 @@
 import * as t from "io-ts/lib/index.js";
 
-// TODO: This should be DELETED as there is a newtype for this in runtime-core,
-//       but a preliminary change broke the build with weird type errors
+// TODO: This should be replaced by a Branded Version available in the Runtime Rest Client Package,
```
[the remark this answers is not in the document]
Comment (nhenin) [reply packages/language/core/v1/src/address.ts:3]: yeah totally agree...

File: packages/runtime/client/rest/src/contract/rolesConfigurations.ts (hunk 17 above)
Comment (nhenin) [reply packages/runtime/client/rest/src/contract/rolesConfigurations.ts:24]: https://github.com/input-output-hk/marlowe-ts-sdk/issues/117

File: packages/runtime/client/rest/src/contract/rolesConfigurations.ts (hunk 18 above)
Comment (nhenin) [reply packages/runtime/client/rest/src/contract/rolesConfigurations.ts:53]: https://github.com/input-output-hk/marlowe-ts-sdk/issues/118

File: packages/adapter/src/codec.ts (hunk 1 above)
Comment (bjornkihlberg) [reply packages/adapter/src/codec.ts:27]: @hrajchert It does normally but `Error` breaks it so you have to manage it manually.

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts (hunk 11 above)
Comment (nhenin) [reply packages/runtime/client/rest/src/contract/endpoints/collection.ts:208]: https://github.com/input-output-hk/marlowe-ts-sdk/issues/118

File: packages/runtime/client/rest/src/contract/endpoints/collection.ts (hunk 13 above)
Comment (nhenin) [reply packages/runtime/client/rest/src/contract/endpoints/collection.ts:313]: I don't like it...I think it is premature optimization

File: packages/runtime/client/rest/src/index.ts [hunk 22]
```diff
@@ -236,7 +237,7 @@ export function mkRestClient(baseURL: string): RestClient {
     },
     buildCreateContractTx(request) {
       const postContractsRequest = {
-        contract: request.contractOrSourceId,
+        contract: "contract" in request ? request.contract : request.sourceId,
```
Comment (bjornkihlberg) [inline packages/runtime/client/rest/src/index.ts:240]: nice and clean

**Pull request merged**